RD-3438 Show images in blueprint

leon-good-life opened the pull request:
<!--- Provide JIRA issue ID and a general summary of your changes in the Title above -->
<!--- e.g. "RD-1234 Improve styling in SuperComponent" -->
<!--- Note that the title is used as a commit message after merging the PR -->

## Description
Jira ticket: https://cloudifysource.atlassian.net/browse/RD-3438
<!--- Describe your changes to help reviewers understand the details. -->

## Screenshots / Videos


https://user-images.githubusercontent.com/17104212/183984564-e9ddf15c-7d0b-4c0e-9e55-a3b7e500fec3.mov



<!--- If applicable, add screenshot(s) or video(s) describing the changes you made. -->
<!--- Consider adding comparison screenshot(s) - before and after the change. -->
<!--- If not applicable, just write “N/A” in this section. -->

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] My code follows the [UI Code Style](https://cloudifysource.atlassian.net/l/c/x8fq902p).
- [x] I verified that all tests and checks have passed.
- [x] I verified if that change requires any update in the [documentation](https://cloudifysource.atlassian.net/l/c/4XKtPocy).
- [x] I added proper labels to this PR.

## Tests
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/3396/pipeline
<!--- If applicable, describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If applicable, provide a link to system tests, e.g: -->
<!--- [#1039](https://jenkins.cloudify.co/.../1039) (2021-08-06 07:05:50) --> 
<!--- If not applicable, just write “N/A” in this section. -->

## Documentation
There nothing to add to Blueprint Sources Widget: https://docs.cloudify.co/latest/working_with/console/widgets/blueprintsources/

<!--- If applicable, describe how you documented or plan to document your changes. -->
<!--- Check UI documentation guidelines - https://cloudifysource.atlassian.net/l/c/EYWJ0XfB - for details. -->
<!--- Provide links to JIRA issues, other PRs or related documentation pages. -->
<!--- If not applicable, just write “N/A” in this section. -->

leon-good-life commented on backend/handler/SourceHandler.ts:
```diff
@@ -116,7 +116,19 @@ export async function browseArchiveFile(req, timestamp, path) {
+    if (mimeType.startsWith('text/')) {
+        return fs.readFile(absolutePath, 'utf-8');
+    }
+    return fs.readFile(absolutePath, '');
```
`return fs.readFile(absolutePath, null);` doesn't work because of `types` library that adds types to nodejs.

kubama commented on backend/handler/SourceHandler.ts:
```diff
@@ -116,7 +116,19 @@ export async function browseArchiveFile(req, timestamp, path) {
+export function getMimeType(req, timestamp, path) {
+    const { blueprintId } = req.params;
+    const absolutePath = pathlib.resolve(browseSourcesDir, `${blueprintId}${timestamp}`, blueprintExtractDir, path);
+    const execSync = require('child_process').execSync;
```
Why not `import`?

leon-good-life replied:
I addressed it: 8c97ac9

kubama commented on backend/handler/SourceHandler.ts:
```diff
@@ -116,7 +116,19 @@ export async function browseArchiveFile(req, timestamp, path) {
+    const { blueprintId } = req.params;
+    const absolutePath = pathlib.resolve(browseSourcesDir, `${blueprintId}${timestamp}`, blueprintExtractDir, path);
+    const execSync = require('child_process').execSync;
+    const mimeType = execSync('file --mime-type -b "' + absolutePath + '"').toString();
```
Personally I would avoid relaying on external dependencies like `file` utility. In this case I would use something like https://www.npmjs.com/package/mime-types to have this functionality self-contained.

leon-good-life replied:
I addressed it: [2927a23](https://github.com/cloudify-cosmo/cloudify-stage/pull/2022/commits/2927a23918cd90e2e6f31cd06a59235c319a3eff)

kubama reviewed (commented):
1. I think it would look better if the image was displayed centered in the view canvas.
2. I think system test for the widget should be extended to cover the new use case.

qooban commented:
I'd like to opt out from the review. I'm sure @kubama will help you in finalizing RD-3438.

As I mentioned last week - take a look at RD-5025 as it may help solving this ticket.

leon-good-life commented:
> I think it would look better if the image was displayed centered in the view canvas.

I will ask the PM - how it should look. 
@ (Ilan Yaniv (can't find his GitHub account)) and  @MichaelShn

kubama commented:
> > I think it would look better if the image was displayed centered in the view canvas.
> 
> I will ask the PM - how it should look. @ (Ilan Yaniv (can't find his GitHub account)) and @MichaelShn

I think it's quite common to use centered view. For example in WebStorm:
![image](https://user-images.githubusercontent.com/49753069/183399369-dffa6c11-abb6-4257-b1e0-b47bd922de49.png)

leon-good-life commented:
I centered the image. 
Here is the commit: [a10691a](https://github.com/cloudify-cosmo/cloudify-stage/pull/2022/commits/a10691ae2a8913ce6fc2e2427d0f9d470f8752e6). 
Here is a video:
https://user-images.githubusercontent.com/17104212/183407770-59b860ad-302f-48b1-8e4d-62f673b5b160.mov

leon-good-life commented:
I'm working on implementing a system test

alexmolev commented:
@leon-good-life please adjust the icon to the middle of the window as @kubama has suggested

leon-good-life commented on widgets/blueprintSources/src/BlueprintSources.tsx:
```diff
@@ -71,12 +72,21 @@ export default function BlueprintSources({ data, toolbox, widget }: BlueprintSou
         const path = selectedKeys[0];
 
+        clearImageUrl();
+        if (path.match(/.(jpg|jpeg|png|gif)$/i)) {
+            setImageUrl(`/console/source/browse/${path}`);
+        }
```
```suggestion
        const isImage = path.match(/.(jpg|jpeg|png|gif)$/i)
        if (isImage) {
            setImageUrl(`/console/source/browse/${path}`);
        }
```

leon-good-life replied:
[9062871](https://github.com/cloudify-cosmo/cloudify-stage/pull/2022/commits/90628711255a3cccdfc98cee2f0d0d79b89c45e7)

leon-good-life commented on widgets/blueprintSources/src/BlueprintSources.tsx:
```diff
@@ -71,12 +72,21 @@ export default function BlueprintSources({ data, toolbox, widget }: BlueprintSou
             .doGetFileContent(path)
-            .then(setContent)
+            .then((content: string) => {
+                if (!path.match(/.(jpg|jpeg|png|gif)$/i)) {
+                    setContent(content);
+                }
```
```suggestion
                const isImage = !path.match(/.(jpg|jpeg|png|gif)$/i)
                if (isImage) {
                    setContent(content);
                }
```

leon-good-life replied:
[9062871](https://github.com/cloudify-cosmo/cloudify-stage/pull/2022/commits/90628711255a3cccdfc98cee2f0d0d79b89c45e7)

leon-good-life commented:
I added a test to `bleprint_sources_spec`, I run it on Jenkins: https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/3415/pipeline

I fixed eslint error and refactored the code.

leon-good-life commented on widgets/blueprintSources/src/BlueprintSources.tsx:
```diff
@@ -36,6 +36,53 @@ interface BlueprintTree {
+        <div className="verticalCenter centeredIcon">
+            <Icon name="file outline" size="big" color="grey" />
+        </div>
+    );
```
I tried changing to semantic-ui while addressing this review comment. But my attempts didn't work. Then I replaced 2 `div`s to `styled-components`: [2e7b5c1](https://github.com/cloudify-cosmo/cloudify-stage/pull/2022/commits/2e7b5c186b60bbe0170210cc9d6b146631984ddb)

I looked up for class names to see if I need to remove the definition from `app/styles/style.scss`:
* `.verticalCenter` is defined in `app/styles/style.scss` but it also used in `BlueprintInfo.tsx`.
* `.centeredIcon` is defined in `widgets/blueprintSources/src/widget.css`
* `.alignHighlight ` is defined in `widgets/blueprintSources/src/BlueprintSources.tsx` I didn't change this `div` to `styled-components` as I was not sure if styles that defined inside the widget should be changed or not.

kubama replied:
> * `.alignHighlight ` is defined in `widgets/blueprintSources/src/BlueprintSources.tsx` I didn't change this `div` to `styled-components` as I was not sure if styles that defined inside the widget should be changed or not.

I would say yes, we also have https://cloudifysource.atlassian.net/wiki/spaces/RD/pages/2148073473/Ongoing+Refactors+in+Cloudify+UI#Moving-CSS-styling-closer-to-the-component.

leon-good-life replied:
I addressed your last comment now, I moved the styles to the component: [436b214](https://github.com/cloudify-cosmo/cloudify-stage/pull/2022/commits/436b214b8e571eabdca79eca1ed50c06fcedaabd)

leon-good-life commented:
I started `blueprint_sources_spec` on Jenkins: https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/3416/pipeline

leon-good-life commented:
@kubama  I fixed backend test. sorry for clicking many times during the day "request a review" 🙂

leon-good-life commented:
I started `blueprint_source_spec`: https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/3428/pipeline

leon-good-life commented:
The test passed

kubama commented on widgets/blueprintSources/src/BlueprintSources.tsx:
```diff
@@ -36,11 +37,29 @@ interface BlueprintTree {
+`;
+
+const AlignHighlight = styled.div`
+    & > .label {
+        cursor: pointer;
+    }
+
+    & > pre {
+        margin-top: 2rem;
+        margin-bottom: 0rem;
+    }
```
Can't we set this styling directly on `HighlightText` and `Label` components?

leon-good-life replied:
I addressed it: [ca562c1](https://github.com/cloudify-cosmo/cloudify-stage/pull/2022/commits/ca562c1f330822c264f92d7bc289a9074a3f43d0) and [d4b92e3](https://github.com/cloudify-cosmo/cloudify-stage/pull/2022/commits/d4b92e37250348fe0e0eddb02107789c0fdc0fd0)

`HighlightText` is not `pre` element, `pre` element is a child of `HighlightText`.

kubama replied:
But is the `pre` element being rendered as the root element for this component?

leon-good-life replied:
you are right, I removed `& > pre`: [bc3fef9](https://github.com/cloudify-cosmo/cloudify-stage/pull/2022/commits/bc3fef9f1b1ddabb64953fda4c6f4531c964e92f)

leon-good-life commented:
I'm testing highlight locally right now. I will update.

leon-good-life commented:
I finished testing locally. I'm waiting for Jenkins.

leon-good-life commented:
`blueprint_sources_spec` on Jenkins: https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/3430/pipeline

leon-good-life commented:
The test is passed on Jenkins

kubama commented on widgets/blueprintSources/src/BlueprintSources.tsx:
```diff
@@ -76,22 +70,22 @@ const RightPane = ({ imageUrl, content, filename, type, maximize, isMaximized, m
                     <Modal.Header>{filename}</Modal.Header>
                     <Modal.Content>
-                        <HighlightText language={type}>{content}</HighlightText>
+                        <StyledHighlightText language={type}>{content}</StyledHighlightText>
```
I don't think this change is needed, the original solution didn't style this component inside the modal due to `>` usage, right?

kubama replied:
It's not what I mean.
`HighlightText` is used in two places, on top level and inside the modal. In the original solution it was styled only on the top level. After your change it is being styled in both places. So, I think the styling of `HighlightText` inside `Modal.Content` should be removed to stick to the original styling.

leon-good-life replied:
oh, now i understand. I removed the margins from the modal by using `HighlightText`: [1fb4ae2](https://github.com/cloudify-cosmo/cloudify-stage/pull/2022/commits/1fb4ae22568d60b587fc4b7dfd057c79cdaa2d28)

kubama commented on backend/routes/SourceBrowser.ts:
```diff
@@ -21,7 +21,7 @@ router.get('/browse/:blueprintId/file/:timestamp/*', (req, res, next) => {
         const mimeType = getMimeType(req, timestamp, path);
         browseArchiveFile(req, timestamp, path)
-            .then(content => res.contentType(mimeType).send(content))
+            .then(content => res.contentType('text/plain').send(content))
```
After this change the `mimeType` local variable is no longer in use.

leon-good-life replied:
ye, I troubleshooted a bug and forgot to put it back.

I put bag mimeType: [bc728a9](https://github.com/cloudify-cosmo/cloudify-stage/pull/2022/commits/bc728a9fac25f5612a24ecb93b81a8d94033446c)

leon-good-life commented on backend/handler/SourceHandler.ts:
```diff
@@ -125,9 +125,14 @@ export async function browseArchiveFile(req, timestamp, path) {
+        const mimeType = mime.lookup(absolutePath);
+        return mimeType;
+    } catch (error) {
+        return 'text/plain';
```
I tried to fix a bug error when trying to open `.pyc` files. 
I reverted it, and solved the bug another way: [bc728a9](https://github.com/cloudify-cosmo/cloudify-stage/pull/2022/commits/bc728a9fac25f5612a24ecb93b81a8d94033446c)

leon-good-life commented:
Updated demo video:

https://user-images.githubusercontent.com/17104212/183984564-e9ddf15c-7d0b-4c0e-9e55-a3b7e500fec3.mov
